drv/cosmo-seq-server/src/main.rs

```diff
@@ -0,0 +1,522 @@
+    fn log_pg_registers(&self) {
+        ringbuf_entry!(Trace::RegPgValues {
+            rail_pgs: (&self.seq.rail_pgs).into(),
+            rail_pgs_max_hold: (&self.seq.rail_pgs_max_hold).into(),
+        });
+    }
+
+    fn set_state_impl(
+        &mut self,
+        state: PowerState,
+        why: StateChangeReason,
+    ) -> Result<(), CpuSeqError> {
+        let now = sys_get_timer().now;
+        ringbuf_entry!(Trace::SetState {
+            prev: Some(self.state),
+            next: state,
+            why,
+            now,
+        });
+
+        use fmc_periph::A0Sm;
+        match (self.get_state_impl(), state) {
+            (PowerState::A2, PowerState::A0) => {
+                self.seq.power_ctrl.modify(|m| m.set_a0_en(true));
+                let mut okay = false;
+                // Wait 2 seconds for power-up
+                for _ in 0..200 {
+                    let state = self.log_state_registers();
+                    match state.seq {
+                        Ok(A0Sm::Done) => {
+                            okay = true;
+                            break;
+                        }
+                        Ok(A0Sm::Faulted) | Err(_) => {
+                            break;
+                        }
+                        _ => (),
+                    }
+                    hl::sleep_for(10);
+                }
+
+                if !okay {
+                    // We'll return to A2, leaving jefe and our local state
+                    // unchanged (since they're set after this block).
+                    self.log_pg_registers();
+                    self.seq.power_ctrl.modify(|m| m.set_a0_en(false));
+
+                    // XXX faulted isn't strictly a timeout, but this is the
+                    // closest available error code
+                    return Err(CpuSeqError::A0Timeout);
+                }
+
+                // Flip the host flash mux so the CPU can read from it
+                // (this is secretly infallible on Cosmo, so we can unwrap it)
+                self.hf.set_mux(drv_hf_api::HfMuxState::HostCPU).unwrap();
+            }
+            (PowerState::A0, PowerState::A2)
+            | (PowerState::A0PlusHP, PowerState::A2)
+            | (PowerState::A0Thermtrip, PowerState::A2) => {
+                self.seq.power_ctrl.modify(|m| m.set_a0_en(false));
+                let mut okay = false;
+                for _ in 0..200 {
+                    let state = self.log_state_registers();
+                    match state.seq {
+                        Ok(A0Sm::Idle) => {
+                            okay = true;
+                            break;
+                        }
+                        Ok(A0Sm::Faulted) | Err(_) => {
+                            break;
+                        }
+                        _ => (),
+                    }
+                    hl::sleep_for(10);
+                }
+                if !okay {
+                    self.log_pg_registers();
```

Comment:
logging errors here, including nic states is probably helpful as well, same as comment above.